Automate Control Plane RG Setup (#480)
microsoft/MLOS: scripts/setup-rg/README.md

Reviewer: Proposal:
Let's reorg all of these into the following structure:

```
mlos_core/mlos_core/config/services/remote/azure/scripts/setup-rg/...
```

A convenience could also be to add a symlink:

```
scripts/azure/setup-rg -> ../../mlos_bench/mlos_bench/config/services/remote/azure/scripts/setup-rg/
```

Reviewer: Should also add a link to this README.md from one of the other ones (e.g. `mlos_bench/mlos_bench/README.md`)

Reviewer 2: [services/remote/azure/README.md](../../blob/main/mlos_bench/mlos_bench/services/remote/azure/README.md) refers to this README - can you add a link to this file URL instead of just mentioning a file name?

Also, maybe describe it at [mlos_bench/README.md](../../blob/main/mlos_bench/README.md) because this is where we outline all steps required to run `mlos_bench`

# Resource Group Setup

> These scripts are helpers for setting up a new resource group (RG) as the control plane for MLOS.

Reviewer: ```suggestion
These scripts are helpers for setting up a new resource group (RG) in Azure as the *control plane* for MLOS.

The *control plane RG* is a container for the *persistent* resources of MLOS (results/metrics storage, scheduler VM, notebook interface, etc.).
```

## Quickstart

1. Starting in this current directory, ensure that we are logged in to Azure CLI.

    ```sh
    az login
    ```

2. Make a copy of the ARM parameters file.

    ```sh

>     cp rg-template.example.parameters.json rg-template.<your name>.parameters.json

Reviewer: ```suggestion
    cp rg-template.example.parameters.json rg-template.parameters.json
```

    ```

3. Modify the ARM parameters in the newly created file as needed, especially the `PLACEHOLDER` values.

> 4. Execute the main script and follow prompts:

Reviewer: Let's not give directions that rely on prompts.  Only CLI args.  Thanks!

    ```sh
    ./setup-rg.ps1
    Supply values for the following parameters:
    armParameters: rg-template.<your name>.parameters.json
    servicePrincipalName: <name of service principal for generating tokens with>
    resourceGroupName: <target resource group for this control plane setup>
    certName: <desired key vault name of certificate for the service principal e.g. mlos-autotune-sp-cert>
    ```

## Manual

Parameters for script can also passed in manually (without prompt) as follows:

```sh
./setup-rg.ps1 `
    -armParameters $armParams `
    -servicePrincipalName $servicePrincipalName `
    -resourceGroupName $resourceGroupName `
    -certName $certName
```

where `$armParams` follows the same usage as `--parameters` in [az deployment group create](https://learn.microsoft.com/en-us/cli/azure/deployment/group?view=azure-cli-latest#az-deployment-group-create-examples).

## Workflow

The high-level flow for what this script automates is as follows:

> 1. Assign `Contributor` access to the Service Principal (SP) for write access over resources.

Reviewer: Does it optionally create the SP?

Author: Now optionally creates the SP!

>     For now we assume the experiment's resources are provisioned in the same RG as the control plane, so the access is granted to the same RG.

Reviewer: We probably need a PR for the control plane setup too.

>     For now we assume the experiment's resources are provisioned in the same RG as the control plane, so the access is granted to the same RG.

Reviewer: ```suggestion
    Ideally, experiment resources are placed in their own RG.
    When that isn't possible, they can also be placed in the control plane RG, in which case the SP can optionally be given access to the control plane RG as well.
```

Reviewer: With associated script fixups to make that indeed optional.

2. Provision control plane resources into the RG.
    This includes:

>     - Networking (public IP, security group, vnet, subnet, network interface)

Reviewer: Most of these probably shouldn't be there in the control plane.

Reviewer: Actually, the NSG might be able to be reused across both experiment resources and the control plane.  We'd have to test it.
But not across locations, so still probably better to just keep them separate.

Author: I agree! I meant for these to be solely for the control VM, and experiments will set up their own as needed. Will rephrase this to `Control VM's networking`

>     - Key Vault

Reviewer: ```suggestion
    - Key Vault for storing the SP credentials.
```

>     - Storage (storage account, file share, MySQL Flex server)

Reviewer: ```suggestion
    - Storage (storage account, file share, DB)
```

- DB need not be MySQL, so we should make that optional as well (can certainly default to MySQL though and let it get parameterized).
- Storage needs to be location dependent, and embedded in the name.
  As such, we should make sure this script can be idempotent or split so that we can independently setup storage in multiple locations without it complaining about the DB portion or scheduler VM already existing.

In fact as far as that goes, we could even have multiple scheduler VMs at some point.

Reviewer: Nm, in the templates that gets hard to support.

We can stick with a MySQL Flex DB for now, but let's do the following:

1. Separate out the template so that some of these resources can be deployed, in an appropriate order, separately.
2. Update the script to allow skip deploying the DB in case someone wants to provide an alternative implementation option there.  Our default supported implementation can be MySQL for now.
3. Label it generically as "results DB" instead.

Author: - Split out an ARM template for the results DB that users can optionally provision.
- Kept storage in the original template, but now allows for users to set a list of locations to create one storage for each, with location is appended to the storage name. This was ease the updating of the control vnet's service endpoints too.

Reviewer: On the last point, will that still work if we want to add additional locations after the fact, or do they all needed to be added at the initial creation time?

Author: It will still work after the fact, it will incrementally add the storages!

>     - VM

Reviewer: ```suggestion
    - VM for running the `mlos_bench` scheduler.
```

3. Assign `Key Vault Administrator` access to the current user.
    This allows the current user to retrieve secrets / certificates from the VM once it is set up.
    Ensure to log in as the same user in the VM.

> 4. If not existing yet, generate a certificate with `certName` in the key vault.

Reviewer: Shouldn't the script either allow us to:
- create the SP and store its generated cert in the keyvault, or
- reference an existing SP with provided cert?

I think the middle case someone else created the SP, but then we reset its password/cert and store that in the keyvault could break other uses of the SP potentially.

Author: I agree! Now the flow is:
- If cert name does not exist in the keyvault already, create or update a SP with the access, generate a cert and store in the keyvault.
- Otherwise, create or update a SP with the access only, and verify that the SP's cert is linked with the one in the keyvault already via thumbprint.
